Fix CVE-2026-107705 by applying the 127-byte password limit before revision6Hash.
Source: https://gitlab.freedesktop.org/poppler/poppler/-/commit/60e19a30152560fcccfb2ae7f7559eda560c0160
---
diff --git a/poppler/Decrypt.cc b/poppler/Decrypt.cc
--- a/poppler/Decrypt.cc
+++ b/poppler/Decrypt.cc
@@ -57,16 +57,18 @@ static void sha256(unsigned char *msg, int msgLen, unsigned char *hash);
 static void sha384(unsigned char *msg, int msgLen, unsigned char *hash);
 static void sha512(unsigned char *msg, int msgLen, unsigned char *hash);
 
-static void revision6Hash(const GooString *inputPassword, unsigned char *K, const char *userKey);
+static void revision6Hash(const std::string &inputPassword, unsigned char *K, const char *userKey);
 
 static const unsigned char passwordPad[32] = { 0x28, 0xbf, 0x4e, 0x5e, 0x4e, 0x75, 0x8a, 0x41, 0x64, 0x00, 0x4e, 0x56, 0xff, 0xfa, 0x01, 0x08, 0x2e, 0x2e, 0x00, 0xb6, 0xd0, 0x68, 0x3e, 0x80, 0x2f, 0x0c, 0xa9, 0xfe, 0x64, 0x53, 0x69, 0x7a };
 
+constexpr int kMaxPasswordLength = 127;
+
 //------------------------------------------------------------------------
 // Decrypt
 //------------------------------------------------------------------------
 
-bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *ownerKey, const GooString *userKey, const GooString *ownerEnc, const GooString *userEnc, int permissions, const GooString *fileID, const GooString *ownerPassword,
-                          const GooString *userPassword, unsigned char *fileKey, bool encryptMetadata, bool *ownerPasswordOk)
+bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *ownerKey, const GooString *userKey, const GooString *ownerEnc, const GooString *userEnc, int permissions, const GooString *fileID, const GooString *ownerPasswordA,
+                          const GooString *userPasswordA, unsigned char *fileKey, bool encryptMetadata, bool *ownerPasswordOk)
 {
     DecryptAES256State state;
     unsigned char test[127 + 56], test2[32];
@@ -74,20 +76,18 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
     unsigned char fState[256];
     unsigned char tmpKey[16];
     unsigned char fx, fy;
-    int len, i, j;
+    int i, j;
 
     *ownerPasswordOk = false;
 
     if (encRevision == 5 || encRevision == 6) {
 
         // check the owner password
-        if (ownerPassword) {
+        if (ownerPasswordA) {
             //~ this is supposed to convert the password to UTF-8 using "SASLprep"
-            len = ownerPassword->size();
-            if (len > 127) {
-                len = 127;
-            }
-            memcpy(test, ownerPassword->c_str(), len);
+            const std::string ownerPassword = ownerPasswordA->toStr().substr(0, kMaxPasswordLength);
+            const int len = ownerPassword.size();
+            memcpy(test, ownerPassword.c_str(), len);
             memcpy(test + len, ownerKey->c_str() + 32, 8);
             memcpy(test + len + 8, userKey->c_str(), 48);
             sha256(test, len + 56, test);
@@ -98,7 +98,7 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
             if (!memcmp(test, ownerKey->c_str(), 32)) {
 
                 // compute the file key from the owner password
-                memcpy(test, ownerPassword->c_str(), len);
+                memcpy(test, ownerPassword.c_str(), len);
                 memcpy(test + len, ownerKey->c_str() + 40, 8);
                 memcpy(test + len + 8, userKey->c_str(), 48);
                 sha256(test, len + 56, test);
@@ -121,13 +121,11 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
         }
 
         // check the user password
-        if (userPassword) {
+        if (userPasswordA) {
             //~ this is supposed to convert the password to UTF-8 using "SASLprep"
-            len = userPassword->size();
-            if (len > 127) {
-                len = 127;
-            }
-            memcpy(test, userPassword->c_str(), len);
+            const std::string userPassword = userPasswordA->toStr().substr(0, kMaxPasswordLength);
+            const int len = userPassword.size();
+            memcpy(test, userPassword.c_str(), len);
             memcpy(test + len, userKey->c_str() + 32, 8);
             sha256(test, len + 8, test);
             if (encRevision == 6) {
@@ -138,7 +136,7 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
             if (!memcmp(test, userKey->c_str(), 32)) {
 
                 // compute the file key from the user password
-                memcpy(test, userPassword->c_str(), len);
+                memcpy(test, userPassword.c_str(), len);
                 memcpy(test + len, userKey->c_str() + 40, 8);
                 sha256(test, len + 8, test);
                 if (encRevision == 6) {
@@ -162,13 +160,13 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
         return false;
     }
     // try using the supplied owner password to generate the user password
-    if (ownerPassword) {
-        len = ownerPassword->size();
+    if (ownerPasswordA) {
+        const int len = ownerPasswordA->size();
         if (len < 32) {
-            memcpy(test, ownerPassword->c_str(), len);
+            memcpy(test, ownerPasswordA->c_str(), len);
             memcpy(test + len, passwordPad, 32 - len);
         } else {
-            memcpy(test, ownerPassword->c_str(), 32);
+            memcpy(test, ownerPasswordA->c_str(), 32);
         }
         md5(test, 32, test);
         if (encRevision == 3) {
@@ -205,7 +203,7 @@ bool Decrypt::makeFileKey(int encRevision, int keyLength, const GooString *owner
     }
 
     // try using the supplied user password
-    return makeFileKey2(encRevision, keyLength, ownerKey, userKey, permissions, fileID, userPassword, fileKey, encryptMetadata);
+    return makeFileKey2(encRevision, keyLength, ownerKey, userKey, permissions, fileID, userPasswordA, fileKey, encryptMetadata);
 }
 
 bool Decrypt::makeFileKey2(int encRevision, int keyLength, const GooString *ownerKey, const GooString *userKey, int permissions, const GooString *fileID, const GooString *userPassword, unsigned char *fileKey, bool encryptMetadata)
@@ -215,13 +213,13 @@ bool Decrypt::makeFileKey2(int encRevision, int keyLength, const GooString *owne
     unsigned char fState[256];
     unsigned char tmpKey[16];
     unsigned char fx, fy;
-    int len, i, j;
+    int i, j;
     bool ok;
 
     // generate file key
     buf = static_cast<unsigned char *>(gmalloc(72 + fileID->size()));
     if (userPassword) {
-        len = userPassword->size();
+        const int len = userPassword->size();
         if (len < 32) {
             memcpy(buf, userPassword->c_str(), len);
             memcpy(buf + len, passwordPad, 32 - len);
@@ -237,7 +235,7 @@ bool Decrypt::makeFileKey2(int encRevision, int keyLength, const GooString *owne
     buf[66] = (permissions >> 16) & 0xff;
     buf[67] = (permissions >> 24) & 0xff;
     memcpy(buf + 68, fileID->c_str(), fileID->size());
-    len = 68 + fileID->size();
+    int len = 68 + fileID->size();
     if (!encryptMetadata) {
         buf[len++] = 0xff;
         buf[len++] = 0xff;
@@ -1764,7 +1762,7 @@ static void sha384(unsigned char *msg, int msgLen, unsigned char *hash)
 // Section 7.6.3.3 (Encryption Key algorithm) of ISO/DIS 32000-2
 // Algorithm 2.B:Computing a hash (for revision 6).
 //------------------------------------------------------------------------
-static void revision6Hash(const GooString *inputPassword, unsigned char *K, const char *userKey)
+static void revision6Hash(const std::string &inputPassword, unsigned char *K, const char *userKey)
 {
     unsigned char K1[64 * (127 + 64 + 48)];
     unsigned char E[64 * (127 + 64 + 48)];
@@ -1772,7 +1770,7 @@ static void revision6Hash(const GooString *inputPassword, unsigned char *K, cons
     unsigned char aesKey[16];
     unsigned char BE16byteNumber[16];
 
-    int inputPasswordLength = inputPassword->size();
+    const int inputPasswordLength = inputPassword.size();
     int KLength = 32;
     const int userKeyLength = userKey ? 48 : 0;
     int sequenceLength;
@@ -1783,7 +1781,7 @@ static void revision6Hash(const GooString *inputPassword, unsigned char *K, cons
         sequenceLength = inputPasswordLength + KLength + userKeyLength;
         totalLength = 64 * sequenceLength;
         // a.make the string K1
-        memcpy(K1, inputPassword->c_str(), inputPasswordLength);
+        memcpy(K1, inputPassword.c_str(), inputPasswordLength);
         memcpy(K1 + inputPasswordLength, K, KLength);
         if (userKey) {
             memcpy(K1 + inputPasswordLength + KLength, userKey, userKeyLength);
